overlord/snapstate/snapstate.go

```diff
@@ -1407,13 +1407,85 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []
 	return updateManyFiltered(ctx, st, names, revOpts, userID, nil, flags, "")
 }
 
-// EnforceSnaps installs/updates/removes snaps reported by validationErrorToSolve.
-// validationSets is the list of sets passed by the user and it's used in the
-// final stage to update validation-sets tracking in the state.
-// userID is used for store auth.
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+// ResolveValidationSetsEnforcementError installs and updates snaps reported by validationErrorToSolve.
+func ResolveValidationSetsEnforcementError(ctx context.Context, st *state.State, valErr *snapasserts.ValidationSetsValidationError, pinnedSeqs map[string]int, userID int) ([]*state.TaskSet, []string, error) {
+	if len(valErr.InvalidSnaps) != 0 {
+		invSnaps := make([]string, 0, len(valErr.InvalidSnaps))
+		for invSnap := range valErr.InvalidSnaps {
+			invSnaps = append(invSnaps, invSnap)
+		}
+		return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints that require removing snaps: %s", strutil.Quoted(invSnaps))
```

Comment:
s/enforcement/validation/ ? or enforced validation

Reply:
Changed to "validation set constraints". Thanks